Update Network Policy Secret management to use SDS

youngnick opened the pull request:
This PR migrates Network Policy Secret handling to use SDS instead of inlining the secrets inside NPDS.

This has a few advantages:
- Changes to policy Secrets will no longer require a policy recalculation to take effect
- We can use the existing secret syncing functionality to reduce the scope of access that the Agent requires to Secrets (so that it no longer requires access to all Secrets in the cluster for secret use in NetworkPolicy to work)
- Reducing the size of NPDS updates (by not inlining) should hopefully speed up sending this config, and help with memory use at large scale.


Some background on using secrets in NetworkPolicy:

There are two ways you can use a Secret in NetworkPolicy, for TLS interception (both termination and origination), and for storing the value of a header that you want to match or add.

TLS Interception seems to be more common, but both are currently implemented.

One additional piece of functionality that this PR preserves is the ability to read "Secrets" specified in CiliumNetworkPolicy via files on the filesystem of the Agent. This functionality was built when Cilium was more focused on non-Kubernetes use cases, and should probably be deprecated and then removed at a later date. Removing this will _substantially_ simplify the code.

This PR also includes all the requisite Helm changes to track secret syncing (similarly to Ingress, Gateway API, and BGP config), and updates the connectivity tests in cilium-cli to use all the new values and functionality correctly.

It also updates the existing TLS Interception docs for the changes, and makes them so that they work (the example app `artii.herokuapp.com` is down, this replaces that with `httpbin.org`.)

Fixes: #24020 

```release-note
Cilium now sends TLS Interception and Header manipulation secrets referenced in CiliumNetworkPolicy and CiliumClusterwideNetworkPolicy by reference using SDS, using the same secret synchronization method used for Ingress, Gateway API, and BGP control plane secrets.
```

For reviewers, reviewing by commits will definitely be easier for this one.

Sorry about the size, this turned out to be a _much_ bigger change than anticipated.

youngnick committed:
Add Secret syncing for CiliumNetworkPolicy TLS Secrets

This commit adds synchronization for any Secrets referenced
in CiliumNetworkPolicy and CiliumClusterwideNetworkPolicy
objects. The secrets are copied into the configured
namespace (default `cilium-secrets`), and kept synchronized
by the Cilium Operator.

Note that this applies to both TLS Secrets and Header Rule
value Secrets.

Signed-off-by: Nick Young <nick@isovalent.com>

youngnick committed:
Present synchronized secrets in SDS

This commit changes Network Policy parsing so that, when secret
synchronization for Network Policy is enabled, referenced
Kubernetes Secrets will be sent using references to SDS secrets,
along with ensuring that synchronized secrets are read into
SDS correctly.

When secret sync for Policy is disabled, the UseFullContext fix
option will work to fix a bug as per the fix for #31761. When
secret sync is enabled, this option is ignored, as SDS secrets
are not susceptible to the same issue.

This change also keeps the existing option to read Secret values
referenced in Network Policy TLS Secrets from the filesystem
rather than Kubernetes. In this case, the secret values will
be sent inline in NPDS.

Signed-off-by: Nick Young <nick@isovalent.com>

youngnick committed:
Enable secret sync handling for header secrets

Signed-off-by: Nick Young <nick@isovalent.com>

youngnick committed:
Fix header secret key handling

This commit fixes key handling for Header Kubernetes Secrets.

Before the migration to SDS, Header Kubernetes Secrets would have
the first value copied blindly into an inline Envoy Secret inside
Network Policy Discovery Service. This had the effect of allowing
any key for the Secret.

This commit keeps this contract by ensuring that any Kubernetes
Secret with a single key will have the associated data copied into
the Envoy Secret for SDS.

This is a change to the current behavior where Secrets with only one
value will not be copied into SDS at all.

Signed-off-by: Nick Young <nick@isovalent.com>

youngnick committed:
Update TLS Interception docs

Updates the TLS Interception docs to have the correct details
for enablement, and changes the example to a working one
(artii.herokuapp.com has disappeared, so this switches to using
httpbin.org instead).

Signed-off-by: Nick Young <nick@isovalent.com>

youngnick committed:
Handle Secret Sync being disabled correctly

This updates the Secret handling such that if secret synchronization
is disabled in Helm, then the previous behavior of the agent
reading secret values directly will still work.

Signed-off-by: Nick Young <nick@isovalent.com>

mhofstetter commented on pkg/crypto/certificatemanager/certificate_manager.go:
```diff
@@ -26,12 +27,14 @@ var Cell = cell.Module(
 type SecretManager interface {
-	GetSecrets(ctx context.Context, secret *api.Secret, ns string) (string, map[string][]byte, error)
-	GetSecretString(ctx context.Context, secret *api.Secret, ns string) (string, error)
+	GetSecrets(ctx context.Context, secret *api.Secret, ns string) (string, map[string][]byte, bool, error)
```
nit: loooks like `GetSecrets` is only used internally of the manager - therefore it could be removed from the interface `SecretManager`

youngnick replied:
Done in 2652aedc2473215bbeaa4b502358d315673ed5d8.

mhofstetter commented on pkg/crypto/certificatemanager/certificate_manager.go:
```diff
@@ -51,33 +54,59 @@ func (mc managerConfig) Flags(flags *pflag.FlagSet) {
+		Logger:    logger,
 	}
 
+	m.Logger.Debug("certificate-manager starting")
```
is this log message a leftover from a debug session? 

just mentioning it because it's misleading having this "starting" in the Hive config lifecycle (where the component isn't actually started - just configured). 

Therefore, I'd recommend to either remove the log message (if no longer necessary) or rephrase it (`Initializing certificate-manager`).

youngnick replied:
Removed.

jrajahalme commented on pkg/crypto/certificatemanager/certificate_manager.go:
```diff
@@ -150,19 +182,23 @@ func (m *manager) GetTLSContext(ctx context.Context, tlsCtx *api.TLSContext, ns
+	// TODO(youngnick): Follow up PR that will change this to a deprecation warning once we actually
+	// mark read-from-file as deprecated.
+	m.Logger.WithField("secret", name).Debug("Secret read from file")
+	return ca, public, private, fromFile, nil
```
Would be a bit clearer to return `false` than `fromFile` here.

youngnick replied:
I've updated the name of the boolean to `inlineSecrets` to make the purpose clearer, but this should actually be `true`, since the Secret is being sent inline via NPDS. Updated.

mhofstetter commented on pkg/policy/l4.go:
```diff
@@ -783,32 +789,43 @@ const (
-	default:
-		return nil, fmt.Errorf("invalid TLS direction: %s", direction)
+	} else {
+		log.Debug("Secret being read from Kubernetes", "secret", types.NamespacedName(*tls.Secret))
```
is this else case and the logging necessary? It looks like it's a duplication as the same message is already logged in `(m *manager) GetTLSContext` (https://github.com/cilium/cilium/blob/47556bd245d901b0fd89da7e830f2acbe68c0709/pkg/crypto/certificatemanager/certificate_manager.go#L170)

youngnick replied:
Removed.

mhofstetter commented on pkg/envoy/xds_server.go:
```diff
@@ -1311,19 +1321,43 @@ func getHTTPRule(secretManager certificatemanager.SecretManager, h *api.PortRule
+							}, policySecretsNamespace),
+						})
+					}
+					// Check for header presence in headers
```
is it correct to add both (`cilium.HeaderMatch` & `HeaderMatcher`) in case the value is empty and `!inlineSecrets`? Or should we handle the `// Check for header presence in headers` in an else condition?

youngnick replied:
Refactored this in 2652aedc2473215bbeaa4b502358d315673ed5d8, what's there is now definitely correct, with tests, and verified with the connectivity tests. Trust me on this, or the rather the six hours I spent testing today.

jrajahalme commented on pkg/crypto/certificatemanager/certificate_manager.go:
```diff
@@ -98,12 +116,19 @@ func (m *manager) GetSecrets(ctx context.Context, secret *api.Secret, ns string)
+		// Files read from disk, so bool returnval is true
+		return nsName, secrets, true, nil
 	}
-	secrets, err := m.k8sClient.GetSecrets(ctx, ns, secret.Name)
```
So we no longer read the secrets and inline them no matter what the feature flags say? So after this the new network policy secrets sync feature has to be enabled for the TLS interception to work at all? If so, what is the point of the feature flag?

youngnick replied:
This is now fixed - we read the secrets and inline them if the feature flag is disabled.

jrajahalme commented on pkg/crypto/certificatemanager/certificate_manager.go:
```diff
@@ -150,19 +182,23 @@ func (m *manager) GetTLSContext(ctx context.Context, tlsCtx *api.TLSContext, ns
 func (m *manager) GetSecretString(ctx context.Context, secret *api.Secret, ns string) (string, error) {
-	name, secrets, err := m.GetSecrets(ctx, secret, ns)
+	// TODO: Update this and to handle read-from-file
+	name, secrets, _, err := m.GetSecrets(ctx, secret, ns)
```
It would be better to return an error the secret was not read from a file, as otherwise we will always return an empty secret. Presumably this function is not used if the feature flag is on, but what if the new feature flag is off?

youngnick replied:
I actually updated this while you were reviewing, sorry.

The behavior if secret sync is disabled is now that secrets will be inlined as today. I'd prefer to have a way to replicate the current behavior in case there is something we've missed in the SDS-and-NPDS handling code in Envoy. I don't think it's likely there's a problem, but this lets users move back to the current behavior by changing a flag.

jrajahalme commented on pkg/envoy/xds_server.go:
```diff
@@ -1369,24 +1378,38 @@ var CiliumXDSConfigSource = &envoy_config_core.ConfigSource{
+// Secrets will be sent to Envoy via the old, inline-in-NPDS method, and _not_ via SDS, and so this method will
+// return whatever is in the *policy.TLSContext.
+func toEnvoyOriginatingTLSContext(tls *policy.TLSContext, policySecretsNamespace string, useFullTLSContext bool) *cilium.TLSContext {
+	if !tls.FromFile && policySecretsNamespace != "" {
```
I was expecting the agent to read the secret from it's original namespace and send it over via SDS if policy secret sync namespace is not specified?

youngnick replied:
As said in other comments, disabling secret sync now restores the current inlined behavior, so we're not depending on new(ish) SDS handling code in all cases. I'll need to come back and follow up in the 1.18 timeframe once we see how the SDS handling code goes in real world use.

mhofstetter commented on pkg/envoy/cell.go:
```diff
@@ -327,12 +334,20 @@ func registerSecretSyncer(params syncerParams) error {
 
+	// Policy secret syncing requires notifying the included xDSServer of the configured value,
+	// so it can be correctly used by the code.
+	if params.Config.EnablePolicySecretsSync {
```
Having the Envoy secret sync to set/propagate the config property of the namespace onto the xdsServer and SecretManager seems wrong to me.

This looks like if the Envoy component is the owner of this information - but actually it should be the other way around.

Therefore, I'd love to see that the config properties (`PolicySecretsNamespace` & `EnablePolicySecretsSync`) are defined on the SecretManager and exposed via a getter methods only. Even if the SecretManager might be deprecated and deleted in the future. (At least the namespace configuration would have to survive anyway.)

Envoy SecretSync & xdsServer should depend on the Manager and read the config where necessary. (Even though these dependencies aren't perfect either - at least this marks the Policy component (SecretManager) as the owner of this information)

wdyt?

Note: Once we would have this refactoring - I plan to do a follow up PR so that components can register a xDS secret sync the same way as it's already possible on the Operator Secret sync level. This way we can already get rid of one explicit dependency to the SecretManager (Envoy Secretsync -> SecretManager)

youngnick replied:
I've done this refactoring in f7251f6ccbb93a4555d4d658e069bb46ba6859c2, but I'll be honest, I'm not sure about it. It feels a little cleaner to me to have all the secret sync config in the one place. PTAL and tell me what you think @mhofstetter.

mhofstetter replied:
As written, in addition to moving the config, we should remove the functions `GetPolicySecretSyncNamespace` (unused anyway) & `SetPolicySecretSyncNamespace` from the `xdsServer` too. Also the config property `policySecretSyncNamespace` in the `xdsServer`.

By adding a dependency from the `xdsServer` to the `SecretManager` (the same as we have one from the `SecretSync`) we should be able to retrieve the secret namespace value when passing it on into the function `getNetworkPolicy`.

This way we have the namespace as field only in the `SecretManager`.

> It feels a little cleaner to me to have all the secret sync config in the one place

IMO it only looks like this because we have the secret sync namespaces of ingress & gw-api here as well. But IF these modules would have packages in the Cilium Agent, these config properties wouldn't live in the secretsync module either. Otherwise we would end up that the secretsync package not only keeps all the config values, but also has to configure all the actual modules (like we did here for the policy secret sync). It's the same for the Operator secret sync where we also use some form of registration mechanism.

youngnick replied:
Okay, this is done in 83f02592cba52b9f13326cad4a18cb8e9cd2e5fb.

jrajahalme commented on pkg/envoy/xds_server.go:
```diff
@@ -1369,24 +1378,38 @@ var CiliumXDSConfigSource = &envoy_config_core.ConfigSource{
+	}
+
+	// If we are not using a synchronized secret or are reading from file, useFullTLSContext
+	// matters.
```
IMO we should not inline k8s secrets even when not syncing secrets to a specific namespace. Especially when the k8s API call to read the secret is removed from the certificate manager?

youngnick replied:
I thought about this some more, and then put the inline secret reading back - if the secret sync namespace is not set. That allows users to easily move back to this current behavior in the event something doesn't work correctly. (I think it's very likely that SDS will be a huge improvement in all cases, but I think it's better to have a way back for a bit as well).

I'll add a todo here.

jrajahalme commented on pkg/envoy/xds_server.go:
```diff
@@ -1395,13 +1418,48 @@ func toEnvoyOriginatingTLSContext(tls *policy.TLSContext) *cilium.TLSContext {
+		return fmt.Sprintf("%s/%s", namespacedName.Namespace, namespacedName.Name)
+	}
+	return fmt.Sprintf("%s/%s-%s", policySecretsNamespace, namespacedName.Namespace, namespacedName.Name)
+}
```
IMO this is a bit unfortunate that SDS secrets have different names depending if they were read from the synced namespace of directly. It would be better if we did not expose the secret sync namespace to Envoy at all so that the SDS secret name would only ever contain the original k8s namespace.

youngnick replied:
Hmm, I guess I agree, but we can do that as a later refactor - because it will also involve refactoring the secret reading for Ingress and Gateway.

jrajahalme commented on pkg/policy/l4.go:
```diff
@@ -783,32 +789,43 @@ const (
-	default:
-		return nil, fmt.Errorf("invalid TLS direction: %s", direction)
+	} else {
+		log.Debug("Secret being read from Kubernetes", "secret", types.NamespacedName(*tls.Secret))
```
should use logfield for the secret?

youngnick replied:
Removed this, as @mhofstetter said it's a duplicate log anyway.

mhofstetter commented on pkg/policy/l4.go:
```diff
@@ -783,32 +789,43 @@ const (
 		CertificateChain: public,
 		PrivateKey:       private,
+		FromFile:         inlineSecrets,
+		Secret:           types.NamespacedName(*tls.Secret),
```
question about a specific use-case:

does applying the Cilium specific xDS resource `NPDS` fail (result in a NACK) if the referenced Secret isn't available via SDS? (e.g. in case the k8s Secret isn't applied yet)

If this would fail, and the K8s secret (that then gets synced via SDS) is created afterwards - I'd assume that the user has to trigger a change on the policy to re-trigger a reconciliation of the Policy calculation (and re-application of the xDS resource NPDS).

(Therefore, my hope is that applying NPDS doesn't fail if the  secret isn't available via SDS yet. And above manual interaction isn't necessary at all.)

youngnick replied:
No, it doesn't fail, because NPDS doesn't check SDS until _after_ the config is ACKed. (All the xDS services work this way, and NPDS is no different).

It's one of the big advantages of the SDS approach, actually, that you no longer need to do a policy regeneration to update a Secret.

mhofstetter reviewed (commented):
Thanks for this big improvement @youngnick :tada: 

Some questions and suggestions inline.

(btw. I'm requesting an additional pair of eyes from @jrajahalme on this).

nebril reviewed (approved):
ci-structure LGTM

jrajahalme commented on pkg/envoy/xds_server.go:
```diff
@@ -1320,19 +1321,43 @@ func getHTTPRule(secretManager certificatemanager.SecretManager, h *api.PortRule
+								Name:      hdr.Secret.Name,
+							}, policySecretsNamespace),
+						})
+					}
```
Adding both the HeaderMatch here and the presence match below seems wrong to me. Since we don't know the value of the SDS secret, we'll have to punt the decision on the type of the match (presence only vs value match) to Envoy.

Also, we have this block above:
```
			log.WithError(err).Warning("Failed fetching K8s Secret, header match will fail")
			// Envoy treats an empty exact match value as matching ANY value; adding
			// InvertMatch: true here will cause this rule to NEVER match.
```
So we need to replicate the same "drop on a missing secret" also via SDS. (see https://github.com/cilium/proxy/pull/992)

Maybe better structure would be to check for the secret reference first, e.g., 
```
	if hdr.Secret == nil {
		value, fromFile, err := getSecretString(secretManager, hdr, ns)
		... old code as it was ...
	} else {
		log.Debugf("HeaderMatches: Adding SDS secret match for header %s", hdr.Name)
		headerMatches = append(headerMatches, &cilium.HeaderMatch{
			MismatchAction: mismatch_action,
			Name:           hdr.Name,
			ValueSdsSecret: namespacedNametoSyncedSDSSecretName(types.NamespacedName{
				Namespace: hdr.Secret.Namespace,
				Name:      hdr.Secret.Name,
			}, policySecretsNamespace),
		})
	}
	...
```

youngnick replied:
The problem with the approach as written is that `hdr.Secret` is only `nil` if no Secret is specified at all; that is, if the value is specified directly. If the value is being read from the file system, then `hdr.Secret` contains a `name` and `namespace` that are used to build the path to read from instead.

But I agree that the logic is a bit convoluted, I'll see if I can tidy it up a bit.

youngnick replied:
Done in 2652aedc2473215bbeaa4b502358d315673ed5d8.

jrajahalme reviewed (changes requested):
I'm somewhat confused on what should happen if the secret sync is not enabled for CNP. It looks like neither inlining nor SDS sync would be happening in that case?

squeed commented on operator/pkg/networkpolicy/secretsync.go:
```diff
@@ -0,0 +1,233 @@
+		Name:      obj.GetName(),
+	}
+
+	cnpList := &cilium_api_v2.CiliumNetworkPolicyList{}
```
Will this actually list against the API, or the local informer?

squeed replied:
I have seen deployments with many network policies; does it make sense to cache a secret -> NP mapping?

youngnick replied:
My assumption was this will list using the controller-runtime pull-through cache, but I guess I could be wrong.

I'm also reluctant to make a cache unless we see problems because then we add cache management problems to the usual eventual consistency ones. We use this same approach in Gateway API handling, and I've heard from folks with thousands of HTTPRoutes that haven't noticed excessive delay time in the recalculations.

sayboras replied:
> My assumption was this will list using the controller-runtime pull-through cache, but I guess I could be wrong.

Yes, we go with the default i.e. read from cache, write via API.

chancez commented on pkg/policy/l4.go:
```diff
@@ -783,32 +789,43 @@ const (
 		return nil, nil
 	}
-	ca, public, private, err := policyCtx.GetTLSContext(tls)
+	ca, public, private, inlineSecrets, err := policyCtx.GetTLSContext(tls)
```
It looks like the interface of `GetTLSContext` does not need to change for what you want to achieve, and it almost looks like a lot of extra code and interface updates that could be skipped. Instead, could you check if ` tlsCtx.Secret` is `nil` here instead?

youngnick replied:
Unfortunately no, the Secret name and namespace will still be set even if it's being read from a file (because the fields are still set in that case - as they are set in the original CiliumNetworkPolicy).

The secret-inlining behavior will need to stay around for a while as well, until we're happy that SDS is as scalable as we all hope it is.

youngnick committed:
Simplify getHTTPRule header match logic

This commit adds extra tests for header value handling
in `pkg/envoy/xds_server_test.go`, and updates
`getHTTPRule()` in `pkg/envoy/xds_server.go` to be
much simpler.

It also removes the `inlineSecrets`handling for
GetSecretString, as it was not providing any extra
information over the string value being empty.

It also unexports the `GetSecrets` function from
CertificateManager, as that's now only used inside
`pkg/crypto/certificate_manager.go`.

Signed-off-by: Nick Young <nick@isovalent.com>

youngnick committed:
Migrate policy secretsync flags to CertificateManager

Signed-off-by: Nick Young <nick@isovalent.com>

jrajahalme commented on pkg/envoy/xds_server.go:
```diff
@@ -1259,19 +1272,27 @@ func getHTTPRule(secretManager certificatemanager.SecretManager, h *api.PortRule
+							HeaderMatchSpecifier: &envoy_config_route.HeaderMatcher_PresentMatch{PresentMatch: true},
+						})
+					}
+
```
IMO it is wrong, or at least very confusing to transform a HeaderMatch into two matches in Envoy like above when mismatch action is "fail-on-mismatch".

The old code treated fail-on-mismatch mismatch action separately as it is the default semantic of a regular Envoy HeaderMatcher. If we want to preserve full backwards compatibility when `hdr.Secret == nil` we should retain the old code as is. Alternatively, we could always translate also fail-on-mismatch mismatch action to `cilium.HeaderMatch` as it should behave the same.

jrajahalme replied:
How about this for this part:
```
		// Fetch the secret
		value, err := getSecretString(secretManager, hdr, ns)
		if err != nil {
			log.WithError(err).Warning("Failed fetching K8s Secret, header match will fail")
			// Envoy treats an empty exact match value as matching ANY value; adding
			// InvertMatch: true here will cause this rule to NEVER match.
			headers = append(headers, &envoy_config_route.HeaderMatcher{
				Name: hdr.Name,
				HeaderMatchSpecifier: &envoy_config_route.HeaderMatcher_StringMatch{
					StringMatch: &envoy_type_matcher.StringMatcher{
						MatchPattern: &envoy_type_matcher.StringMatcher_Exact{
							Exact: "",
						},
					},
				},
				InvertMatch: true,
			})
		} else if value != "" {
			// Inline value provided.
			// Header presence and matching (literal) value needed.
			if mismatch_action == cilium.HeaderMatch_FAIL_ON_MISMATCH {
				// fail on mismatch gets converted for regular HeaderMatcher
				headers = append(headers, &envoy_config_route.HeaderMatcher{
					Name: hdr.Name,
					HeaderMatchSpecifier: &envoy_config_route.HeaderMatcher_StringMatch{
						StringMatch: &envoy_type_matcher.StringMatcher{
							MatchPattern: &envoy_type_matcher.StringMatcher_Exact{
								Exact: value,
							},
						},
					},
				})
			} else {
				log.Debugf("HeaderMatches: Adding %s", hdr.Name)
				headerMatches = append(headerMatches, &cilium.HeaderMatch{
					MismatchAction: mismatch_action,
					Name:           hdr.Name,
					Value:          value,
				})
			}
		} else if hdr.Secret == nil {
			// No inline value and no secret.
			// Header presence for FAIL_ON_MISMSTCH or matching empty value otherwise needed.
			if mismatch_action == cilium.HeaderMatch_FAIL_ON_MISMATCH {
				// Only header presence needed
				headers = append(headers, &envoy_config_route.HeaderMatcher{
					Name:                 hdr.Name,
					HeaderMatchSpecifier: &envoy_config_route.HeaderMatcher_PresentMatch{PresentMatch: true},
				})
			} else {
				log.Debugf("HeaderMatches: Adding %s for an empty value", hdr.Name)
				headerMatches = append(headerMatches, &cilium.HeaderMatch{
					MismatchAction: mismatch_action,
					Name:           hdr.Name,
				})
			}
		} else {
			// A secret is set, so we transform to an SDS value.
			// cilium-envoy takes care of treating this as a presence match if the
			// secret exists with an empty value.
			log.Debugf("HeaderMatches: Adding %s because SDS value is required", hdr.Name)
			headerMatches = append(headerMatches, &cilium.HeaderMatch{
				MismatchAction: mismatch_action,
				Name:           hdr.Name,
				ValueSdsSecret: namespacedNametoSyncedSDSSecretName(types.NamespacedName{
					Namespace: hdr.Secret.Namespace,
					Name:      hdr.Secret.Name,
				}, policySecretsNamespace),
			})
		}
```

jrajahalme replied:
Requires this change in the test as well (to remove the double match):
```
-var expectedHeadersPortRuleHeaderMatchSDS = []*envoy_config_route.HeaderMatcher{
-	{
-		Name:                 "VeryImportantHeader",
-		HeaderMatchSpecifier: &envoy_config_route.HeaderMatcher_PresentMatch{PresentMatch: true},
-	},
-}
+var expectedHeadersPortRuleHeaderMatchSDS []*envoy_config_route.HeaderMatcher
```

youngnick replied:
That's _much_ better, thanks! Applied, passes all the tests, including the connectivity ones.

jrajahalme reviewed (changes requested):
IMO the HeaderMatch translation is still not ready, as there is a case where a single HeaderMatch will get translated to both an SDS secret match and a presence match. Suggested change below.

jrajahalme reviewed (approved):
Great improvement, thanks!

sayboras reviewed (approved):
Niceeee

learnitall reviewed (approved):
:rocket: thanks!

derailed reviewed (approved):
@youngnick Nice work!
